internal/controller/apiextensions/claim/reconciler.go

```diff
@@ -359,6 +359,10 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 
 	if meta.WasDeleted(cm) {
 		log = log.WithValues("deletion-timestamp", cm.GetDeletionTimestamp())
+		if len(cm.GetFinalizers()) > 1 {
```

Comment:
I don't think this is how finalizers are to be used here

From the [k8s docs](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/):

> Finalizers are namespaced keys that tell Kubernetes to wait until specific conditions are met before it fully deletes resources marked for deletion. Finalizers alert [controllers](https://kubernetes.io/docs/concepts/architecture/controller/) to clean up resources the deleted object owned.

To me this sounds like the opposite of what we are trying to do here: Do nothing while there is more than one finalizer.

I believe this is an overly complicated solution for what we actually want: Postpone the deletion of XR and XRC while their children are still around (see #3320).

Reply:
This is not intended to be a solution for #3320 - it is intended to be a solution for #3225 and provide a foundation for #3321.
It can easily be extended to support #3320 but that was not it's primary purpose.

This is exactly what finalizers are intended to do - delay the deletion/garbage collection of a resource while a dependency exists on that resource.  Crossplane is the controller for these resources and so it needs to delay it's delete processing while other resources have dependencies on them.

Reply:
I do think this is a bit of a novel use of finalizers, in that I'm not sure it's common for controller code to consider _all_ finalizers as opposed to only considering finalizers it added.

That's not to say we shouldn't proceed with this direction though. Two thoughts:

* Given it's somewhat novel I think an explanatory comment would make sense.
* Would supporting a single special Crossplane finalizer make sense here? i.e. We would not proceed with deletion if a finalizer like `crossplane.io/has-dependents` still exists, but would otherwise ignore other finalizers?

Reply:
> Typically I see finalizers used as a way for a controller to signal that it needs to do something before the API server should complete the deletion.

Exactly.  Now this idea is extended to the context of Crossplane, in it's role as an API server "on top of" the kubernetes API server.  Where the Kubernetes API server is typically managing native objects that have no existence outside of etcd, Crossplane is managing groups of objects that by definition are composed of other Crossplane objects AND objects that have manifestations outside of Kubernetes itself.  Therefore just like Kubernetes garbage collection uses finalizers to keep objects in etcd until "everyone is done with them", Crossplane can use the same concepts to keep it's collections of local/remote objects in existence until there are no more declared dependencies on them.

> Given it's somewhat novel I think an explanatory comment would make sense.

Definitely - I've also thought about maybe a design document just to save the background on the design for posterity in case someone ever needs to know why it is this way.  Not to mention people wanting to understand how it works and how to use it.

> Would supporting a single special Crossplane finalizer make sense here?

I actually did a lot of thinking about this for #3323 and the problem with a single finalizer is you don't know where it came from or when it's ok to delete it.

If we assume the case of an external entitiy declaring a dependency on a Claim or Composite X.  X now has 2 finalizers, it's own plus one for the remote dependency.  (provider-kubernetes does this today with it's dependsOn feature)  Unfortunately maintaining the finalizer only on the Claim/Composite X is not sufficient for the dependency, since even Foreground Cascading Deletion doesn't put a foregroundDeletion finalizer on the "leaf" or "edge" nodes which have no ownerReferences pointing at them.  Any deletion of X will still cause all of the composed/managed resources that make up X to be deleted by the Composite and Managed Reconcilers because there are no additional finalizers on those "child" objects to prevent their deletion.  X will remain with it's extra finalizer but it will have no child resources associate with it, so it's essentially useless.

That's where #3323 comes in.  It is necessary to reconcile the extra finalizers all the way down to the leaf/edge nodes so that deletion of the entire Claim/Composite structure is blocked, not just the Claim/Composite itself.  However, it is not necessary for the specific dependencies on X to be reconciled to the edges.  Only the fact that X has more than one finalizer is important.  So when the claim/composite reconcilers see more than one finalizer on X, they can put a unique finalizer on all of the child composite/composed resources that it is responsible for.  That increases the finalizer count on those objects and those will add their own finalizers to their children, all of the way out to the leaf/edge nodes which provides the desired block against resource deletion.  It also provides traceability of the
finalizers since we can put the object UID in the finalizer name to indicate where the lock originates from (it had better be the same as the ownerReference object, but still).  This can also handle multiple overlapping dependencies on the same resources without concern for finalizer name conflicts.

When the external finalizer is removed from X, the reconciliation sees only 1 finalizer remaining and removes the unique finalizer from it's children, which flows out to the edges.  They now have no finalizers and get deleted as expected.

I originally thought the implementation of #3323 was going to be dependent on #3321 and fairly complex so I kept them separate, which also simplifies the individual problem descriptions.  However once I did the full investigation it turns out that the changes required for #3323 are fairly trivial and decoupled from #3321.  They won't function correctly without #3321, but they are in different places.  So they can be combined into one PR if that's preferred.

Sorry for the novel.  I'll definitely put together a design document.

Reply:
> I've also thought about maybe a design document just to save the background on the design for posterity in case someone ever needs to know why it is this way.

+1 - I think a design doc to tie together the handful of proposals would be very useful.

> I actually did a lot of thinking about this for https://github.com/crossplane/crossplane/issues/3323 and the problem with a single finalizer is you don't know where it came from or when it's ok to delete it.

Yeah, this occurred to me in retrospect as well.

Reply:
Design doc has been added, limited to foreground cascading deletion functionality.